server/http/ipfs.go

```diff
@@ -0,0 +1,195 @@
+			ext := filepath.Ext(filename)
+			if ext == "" {
+				logger.LogStatus(http.StatusBadRequest, "Filename missing extension")
+				res.WriteHeader(http.StatusBadRequest)
+				return
+			}
+			if ext != ".car" {
+				logger.LogStatus(http.StatusBadRequest, fmt.Sprintf("Filename uses non-supported extension %s", ext))
+				res.WriteHeader(http.StatusBadRequest)
+				return
+			}
+		}
+
+		// validate cid path parameter
+		cidStr := urlPath[1]
+		rootCid, err := cid.Parse(cidStr)
+		if err != nil {
+			logger.LogStatus(http.StatusInternalServerError, "Failed to parse cid path parameter")
+			res.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		// Grab unixfs path if it exists
+		// var unixfsPath []string
+		// if len(urlPath) > 2 {
+		// 	unixfsPath = urlPath[2:]
+		// }
+		// TODO: Do something with unixfs path
+
+		log.Debug("creating temp car file")
+		carfile, err := os.CreateTemp("", rootCid.String())
+		if err != nil {
+			logger.LogStatus(http.StatusInternalServerError, fmt.Sprintf("Failed to create temp car file before retrieval: %v", err))
+			res.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		// create blockstore from defined car file and use it for the link system
+		blockstore, err := blockstore.OpenReadWriteFile(carfile, []cid.Cid{rootCid})
+		if err != nil {
+			logger.LogStatus(http.StatusInternalServerError, fmt.Sprintf("Failed to create blockstore from temp car file before retrieval: %v", err))
+			res.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		linkSystem := storeutil.LinkSystemForBlockstore(blockstore)
+
+		log.Debugw("fetching cid into car file", "cid", rootCid.String(), "file", carfile.Name())
+		id, _, err := lassie.Fetch(req.Context(), rootCid, linkSystem)
+		if err != nil {
+			logger.LogStatus(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch cid: %v", err))
+			res.WriteHeader(http.StatusInternalServerError)
+			carfile.Close()
+			return
+		}
+
+		log.Debugw("closing car file", "file", carfile.Name())
+		err = carfile.Close()
+		if err != nil {
+			logger.LogStatus(http.StatusInternalServerError, fmt.Sprintf("Failed to close temp car file after retrieval: %v", err))
+			res.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		// set Content-Disposition header based on filename url parameter
+		var filename string
+		if req.URL.Query().Has("filename") {
+			filename = req.URL.Query().Get("filename")
+		} else {
+			filename = fmt.Sprintf("%s.car", rootCid.String())
+		}
+		res.Header().Set("Content-Disposition", "attachment; filename="+filename)
+
+		res.Header().Set("Accept-Ranges", "none")
+		res.Header().Set("Cache-Control", "public, max-age=29030400, immutable")
+		res.Header().Set("Content-Type", "application/vnd.ipld.car; version=1")
+		res.Header().Set("Etag", fmt.Sprintf("%s.car", rootCid.String()))
+		res.Header().Set("C-Content-Type-Options", "nosniff")
+		res.Header().Set("X-Ipfs-Path", req.URL.Path)
+		// TODO: set X-Ipfs-Roots header
```

Comment:
fwiw these two make sense mostly in deserialized contexts: allow for partial HTTP Cache invalidation based on DAG/path subset that changed (https://github.com/ipfs/kubo/pull/8720, based on request from Cloudflare)

For requests for  `/ipfs/cid` without any sub path they don't bring much value, and  `Etag` is more than enough.

So fine to skip X-Ipfs-Roots header until you support sub-paths (if its even planned)